arc.1
.TH ARC 1 "11 Nov 1991" "Howard Chu@JPL" "LOCAL COMMANDS"
.SH NAME
arc \- pc archive utility
.SH SYNOPSIS
.B arc
a|m|u|f|d|x|e|r|p|l|v|t|c [ biswnoq ] [ g\fIpassword\fR ]
.I archive 
[ \fIfilename\fR ...]
.SH DESCRIPTION
.I Arc
is a general archive and file compression utility, used to maintain
a compressed archive of files.
An
.I archive
is a single file that combines many files, reducing storage space
and allowing multiple files to be handled as one.
.I Arc
uses one of several compression methods for each file within the
.IR archive ,
based on whichever method yields the smallest result.
.SH INSTRUCTIONS
Execute
.I arc
with no arguments for fairly verbose, usable instructions.
.SH COMMAND SWITCHES
.TP 3
a
add files to archive.  Copies the indicated files to the archive.
.TP
m
move files to archive.  Same as 'a' switch except
that the files are deleted from the directory as
they are moved to the archive.
.TP
u
update files in archive.  This switch will replace archived
files when the named file is newer than the archived copy.
New files will be added automatically.
.TP
f
freshen files in archive.  Same as 'u' except that
new files will not be added.
.TP
d
delete files in archive.  The named files are removed from the archive.
.TP
x,e
extract files from archive.  The named files are extracted
from the archive and created in the current directory
in an uncompressed state.
.TP
r
run one file with arguments from archive.  Any
program may be executed directly from the archive.
The parameters given after the program name are passed to
the program without modification.
.TP
p
copy files from archive to standard output.  Useful
with I/O redirection. A form-feed is appended after each file,
to ease use with printers.
.TP
l
list files in archive.  Limited information listing
of files contained in an archive.  Displays the
filename, original length, and date last modified.
If the 'n' option (see below) is used, only the
filename is displayed.
.TP
v
verbose listing of files in archive.  Complete
information listing of files contained in an archive.
Displays the filename, original length, storage method,
storage factor (% savings), compressed size, date, time,
and CRC.
.TP
t
test archive integrity.  Computes CRC values for each member of
the archive and compares against the previously saved value.
.TP
c
convert entry to new packing method.  Convert files
stored with older methods to newer methods that are
more efficient. Also useful for files previously
archived with the 's' option.
.SH OPTIONS
.TP 3
b
retain backup copy of archive.  Keep the original
archive file and rename to .BAK.
This switch may be used with the
following commands:  a, m, u, f, d, c.
.TP
i
suppress image mode.  This switch causes files to
be treated as text files, and will translate their
end-of-line sequence. (Unix's '\\n' vs. '\\r\\n' used
on many other systems.)  The default is to perform
no translation when compressing or extracting files.
This option makes dealing with text files much nicer,
though the 'tr' command can also be used. ('\\r' in
makefiles and C source code is such a nuisance...)
.TP
s
suppress compression.  This forces new files to be
saved using Method 2 (no compression).  This switch
may be used with the following commands:  a, m, u, f, c.
.TP
w
suppress warning messages.  This switch will keep
warning messages from being displayed which is the default.
Most warnings concern the deletion or existence of
files with the same name.
.TP
n
suppress notes and comments.  This switch will keep
useful notes from being displayed which is the default.
Most notes indicate what stage of compression is
being run (analyze, compaction, storage).
.TP
o
overwrite existing files when extracting.  This switch
will make existing files silently get overwritten, instead
of asking for confirmation, which is the default.
.TP
q
force Squash compression method.  This switch causes
the Squash compression method to be used, instead of
Crunch, which is the default.
.TP
g
encrypt/decrypt archive entry.  This is used to encode
files so that others may not read them.  BE CAREFUL!
This must be the last parameter in the switches because
everything following is part of the password.
.SH PROGRAMMING NOTES
.I Arc
Version 2 differs from version 1 in that archive entries
are automatically compressed when they are added to the archive,
making a separate compression step unnecessary.  The nature of the
compression is indicated by the header version number placed in
each archive entry, as follows:
.nf
         1 = Old style, no compression
         2 = New style, no compression
         3 = Compression of repeated characters only
         4 = Compression of repeated characters plus Huffman SQueezing
         5 = Lempel-Zev packing of repeated strings (old style)
         6 = Lempel-Zev packing of repeated strings (new style)
         7 = Lempel-Zev Williams packing with improved hash function
         8 = Dynamic Lempel-Zev packing with adaptive reset
         9 = Squashing
.fi

Type 5, Lempel-Zev packing, was added as of version 4.0

Type 6 is Lempel-Zev packing where runs of repeated characters
have been collapsed, and was added as of version 4.1

Type 7 is a variation of Lempel-Zev using a different hash
function which yields speed improvements of 20-25%, and was
added as of version 4.6

Type 8 is a different implementation of Lempel-Zev, using a
variable code size and an adaptive block reset, and was added
as of version 5.0

Type 9 is another variation of Lempel-Zev, using a larger
hash table. This method was developed by Phil Katz, and is
not supported by the "official" \fBARC\fP programs.

.I Arc
will look for environment variables named \fIARCTEMP\fP or
\fITMPDIR\fP, which, if present, indicates the pathname
where temporary files should be created. This is typically
the location of a RAMdisk on a microcomputer, "/tmp/" or
left unset.

See the included documentation file for more details.
.SH HISTORY
\fIArc\fP has been in use in the CP/M and MSDOS world for many years.
Thom Henderson developed the original version, but it is important to note that
\fIarc\fP is based on the file compression theories developed by Huffman, Welch,
Knott, Knuth, and many other scientists. This implementation is based on
version 5.21 of the MSDOS program.
.SH BUGS
\fIArc\fP behaves just like the PC version of the program; all functions
of the "usage" display are working.
Full compatibility with PC ARC files is maintained, the price for which is
that \fIarc\fP doesn't like long filenames, and can only archive files with
names of up to 12 characters.
It will *sometimes* do The Right Thing with them, but I suggest
you put long-winded filenames in a "shar" before
.IR arc ing
them.

There shouldn't be any problems, (hah!) but if you find any, please
send them to me at:

	hyc@hanauma.jpl.nasa.gov

.SH AUTHORS
Original MSDOS program by Thom Henderson
.br
COPYRIGHT(C) 1985-87 by System Enhancement Associates; ALL RIGHTS RESERVED

Original Lempel-Zev code derived from compress 4.0.
Modified to support Squashing by Dan Lanciani (ddl@harvard.edu)
Ported from MSDOS by Howard Chu,
with help from John Gilmore (hoptoad!gnu), James Turner (daisy!turner)
and others.


Changelog
Unreleased changes:

Klaus Reimer <k@ailis.de>, 01 Jan 2011:
  * Corrected spelling error in arc man page.
  * Version in binaries now comes from VERSION variable in Makefile.
  * Changed man page sections from 1L to 1.


                        CHANGES IN VERSION 5.21p
                 (Klaus Reimer <k@ailis.de>, 07 Aug 2010)

Replaced mktemp with mkstemp.
Set correct version number and updated creation date in command-line help.
Corrected content of PATCHLEVEL file.
Added man page for marc.
Added install make target.
Added -Wall compiler flag.


                        CHANGES IN VERSION 5.21o
                 (Howard Chu <hyc@highlandsun.com>, 12 Oct 2005)

Silence warnings about unused results.


                        CHANGES IN VERSION 5.21n
                 (Howard Chu <hyc@highlandsun.com>, 08 Oct 2005)

Better fix for temporary file creations, previous patch did not fix marc.
Silence all gcc-4.0.0 -Wall warnings.


                        CHANGES IN VERSION 5.21m
                 (Klaus Reimer <k@ailis.de>, 21 Sep 2005)

Applied patch from Joey Schulze fixing the following problems:
  
  - Fixed insecure permissions of a temporary file [arcsvc.c, CAN-2005-2945]
  - Fixed insecure temporary file creation [arcsvc.c, CAN-2005-2992]



                        CHANGES IN VERSION 5.21l
                 (Klaus Reimer <k@ailis.de>, 13 Mar 2005)
                        
Fixed warnings because missing inclusion of string.h and fixed a gcc-4.0
problem with marc.c (Missing declaration of makefnam).



                        CHANGES IN VERSION 5.21k
                 (Klaus Reimer <k@ailis.de>, 6 Nov 2004)
                        
Ulf Harnhammar found some buffer overflows causing segfaults and provided a
patch which is now applied to arc to close these bugs.



                        CHANGES IN VERSION 5.21


The following changes have been made in ARC version 5.21, but have not
yet been included in the manual:

o   When adding files to an archive, ARC now shows the stowage facter
    achieved on each file.

o   A bug was found that would keep an archive entry from being
    encrypted if it was stored without compression.  This has now been
    fixed.

o   If changes are made to a corrupted archive, the corrupted entries
    are discarded.  This makes it possible to lose data accidentally.
    ARC will now only make changes to a corrupted archive if the W
    (suppress Warnings) option has been given.

o   The N (suppress Notes) option now suppresses the "Creating
    archive" note when a new archive is being created.

o   The N (suppress Notes) option formerly did nothing useful when
    used with the L (List files) command.  It now causes a terse
    listing of filenames only, suitable for use with pipes and
    redirection.

o   The list of filenames given to ARC may now include indirect
    references.  If a filename begins with an "at sign" ("@"), it is
    taken to be the name of a file which contains a list of file
    names.  The list of file names may include further indirection.
    If no extension is given, ".CMD" is assumed.  For example, the
    command:

         arc a waste junk.txt @trash

    would cause ARC to add JUNK.TXT plus all files listed in the file
    TRASH.CMD to an archive named WASTE.ARC.  If no file is specified,
    then the list is read from standard input.  For example, the
    command:

         arc ln waste | arc a trash @

    would cause ARC to add files to TRASH.ARC based on the names of
    the files stored in WASTE.ARC.  It is probably a good idea to give
    the O (Overwrite) option if you are extracting files this way.

    Version 5.21 of MARC also allows for indirection.


arc.c
/*
 * $Header: /cvsroot/arc/arc/arc.c,v 1.7 2011/01/01 13:02:30 k_reimer Exp $
 */

/*  ARC - Archive utility
  
    Version 5.21, created on 04/22/87 at 15:05:21
  
(C) COPYRIGHT 1985-87 by System Enhancement Associates; ALL RIGHTS RESERVED
  
    By:	 Thom Henderson
  
    Description:
	 This program is a general archive utility, and is used to maintain
	 an archive of files.  An "archive" is a single file that combines
	 many files, reducing storage space and allowing multiple files to
	 be handled as one.
  
    Instructions:
	 Run this program with no arguments for complete instructions.
  
    Programming notes:
	 ARC Version 2 differs from version 1 in that archive entries
	 are automatically compressed when they are added to the archive,
	 making a separate compression step unecessary.	 The nature of the
	 compression is indicated by the header version number placed in
	 each archive entry, as follows:
  
	 1 = Old style, no compression
	 2 = New style, no compression
	 3 = Compression of repeated characters only
	 4 = Compression of repeated characters plus Huffman SQueezing
	 5 = Lempel-Zev packing of repeated strings (old style)
	 6 = Lempel-Zev packing of repeated strings (new style)
	 7 = Lempel-Zev Williams packing with improved hash function
	 8 = Dynamic Lempel-Zev packing with adaptive reset
	 9 = Dynamic Lempel-Zev packing, larger hash table
  
	 Type 5, Lempel-Zev packing, was added as of version 4.0
  
	 Type 6 is Lempel-Zev packing where runs of repeated characters
	 have been collapsed, and was added as of version 4.1
  
	 Type 7 is a variation of Lempel-Zev using a different hash
	 function which yields speed improvements of 20-25%, and was
	 added as of version 4.6
  
	 Type 8 is a different implementation of Lempel-Zev, using a
	 variable code size and an adaptive block reset, and was added
	 as of version 5.0
  
	 Type 9 is a slight modification of type 8, first used by Phil
	 Katz in his PKARC utilites. The primary difference is the use
	 of a hash table twice as large as for type 8, and that this
	 algorithm called Squashing, doesn't perform run-length encoding
	 on the input data.
  
	 Verion 4.3 introduced a temporary file for holding the result
	 of the first crunch pass, thus speeding up crunching.
  
	 Version 4.4 introduced the ARCTEMP environment string, so that
	 the temporary crunch file may be placed on a ramdisk.	Also
	 added was the distinction bewteen Adding a file in all cases,
	 and Updating a file only if the disk file is newer than the
	 corresponding archive entry.
  
	 The compression method to use is determined when the file is
	 added, based on whichever method yields the smallest result.
  
    Language:
	 Computer Innovations Optimizing C86
*/
#include <stdio.h>
#include "arc.h"

#if	UNIX
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <errno.h>
#endif

#include <string.h>
#if BSD
#include <strings.h>
#endif

#if	!__STDC__
char		*calloc(), *malloc(), *realloc();
#endif

VOID		addarc(), delarc(), extarc(), lstarc(), tstarc(), cvtarc(), runarc();
VOID		arcdie();
static	VOID	expandlst();
#if	_MTS
VOID		etoa();
#endif
#if	GEMDOS
long		_stksize = 30720;
#endif
char		*makefnam();	/* filename fixup routine */

	/* Block I/O buffers */

u_char		*pinbuf, *pakbuf, *outbuf, *outend;

static char   **lst;		/* files list */
static int	lnum;		/* length of files list */

int
main(num, arg)			/* system entry point */
	int		num;	/* number of arguments */
	char	       *arg[];	/* pointers to arguments */
{
	char		opt = 0;/* selected action */
	char	       *a;	/* option pointer */
	VOID		upper();/* case conversion routine */
	char	       *envfind();	/* environment searcher */
	int		n;	/* index */
	char	       *arctemp2;
#if	GEMDOS
	VOID		exitpause();
	int		append;
#endif
#if	_MTS
	fortran VOID	guinfo();
	char		gotinf[4];
#endif
#if	UNIX
	struct	stat	sbuf;
#endif

	if (num < 3) {
		printf("ARC - Archive utility, Version %s, created on 08/07/2010\n", VERSION);
/*		printf("(C) COPYRIGHT 1985,86,87 by System Enhancement Associates;");
		printf(" ALL RIGHTS RESERVED\n\n");
		printf("Please refer all inquiries to:\n\n");
		printf("       System Enhancement Associates\n");
		printf("       21 New Street, Wayne NJ 07470\n\n");
		printf("You may copy and distribute this program freely,");
		printf(" provided that:\n");
		printf("    1)	 No fee is charged for such copying and");
		printf(" distribution, and\n");
		printf("    2)	 It is distributed ONLY in its original,");
		printf(" unmodified state.\n\n");
		printf("If you like this program, and find it of use, then your");
		printf(" contribution will\n");
		printf("be appreciated.	 You may not use this product in a");
		printf(" commercial environment\n");
		printf("or a governmental organization without paying a license");
		printf(" fee of $35.  Site\n");
		printf("licenses and commercial distribution licenses are");
		printf(" available.  A program\n");
		printf("disk and printed documentation are available for $50.\n");
		printf("\nIf you fail to abide by the terms of this license, ");
		printf(" then your conscience\n");
		printf("will haunt you for the rest of your life.\n\n"); */
#if	MSDOS
		printf("Usage: ARC {amufdxerplvtc}[bswnoq][g<password>]");
#endif
#if	GEMDOS
		printf("Usage: ARC {amufdxerplvtc}[bhswnoq][g<password>]");
#endif
#if	UNIX
		printf("Usage: arc {amufdxerplvtc}[biswnoq][g<password>]");
#endif
#if	_MTS
		printf("Parameters: {amufdxeplvtc}[biswnoq][g<password>]");
#endif
		printf(" <archive> [<filename> . . .]\n");
		printf("Where:	 a   = add files to archive\n");
		printf("	 m   = move files to archive\n");
		printf("	 u   = update files in archive\n");
		printf("	 f   = freshen files in archive\n");
		printf("	 d   = delete files from archive\n");
		printf("	 x,e = extract files from archive\n");
#if	!_MTS
		printf("	 r   = run files from archive\n");
#endif
		printf("	 p   = copy files from archive to");
		printf(" standard output\n");
		printf("	 l   = list files in archive\n");
		printf("	 v   = verbose listing of files in archive\n");
		printf("	 t   = test archive integrity\n");
		printf("	 c   = convert entry to new packing method\n");
		printf("	 b   = retain backup copy of archive\n");
#if	GEMDOS
		printf("	 h   = hold screen after finishing\n");
#endif
#if	_MTS
		printf("	 i   = suppress ASCII/EBCDIC translation\n");
#endif
#if	UNIX
		printf("	 i   = suppress image mode (translate EOL)\n");
#endif
		printf("	 s   = suppress compression (store only)\n");
		printf("	 w   = suppress warning messages\n");
		printf("	 n   = suppress notes and comments\n");
		printf("	 o   = overwrite existing files when");
		printf(" extracting\n");
		printf("	 q   = squash instead of crunching\n");
		printf("	 g   = Encrypt/decrypt archive entry\n");
		printf("\nAdapted from MSDOS by Howard Chu\n");
		/*
		 * printf("\nPlease refer to the program documentation for");
		 * printf(" complete instructions.\n"); 
		 */
#if	GEMDOS
		exitpause();
#endif
		return 1;
	}
	/* see where temp files go */
#if	!_MTS
	arctemp = calloc(1, STRLEN);
	if (!(arctemp2 = envfind("ARCTEMP")))
		arctemp2 = envfind("TMPDIR");
	if (arctemp2) {
		strncpy(arctemp, arctemp2, STRLEN - 16);
		arctemp[STRLEN - 17] = '\0';
		n = strlen(arctemp);
		if (arctemp[n - 1] != CUTOFF)
			arctemp[n] = CUTOFF;
	}
#if	UNIX
	else	strcpy(arctemp, "/tmp/");
#endif
#if	!MSDOS
	{
		strcat(arctemp, "AXXXXXX");
		int fd = mkstemp(arctemp);
		if (fd == -1)
		{
			fprintf(stderr, "Unable to create temporary files: %s\n", strerror(errno));
			exit(235);
		}
		close(fd);
		unlink(arctemp);
	}
#else
	strcat(arctemp, "$ARCTEMP");
#endif
#else
	guinfo("SHFSEP	", gotinf);
	sepchr[0] = gotinf[0];
	guinfo("SCRFCHAR", gotinf);
	tmpchr[0] = gotinf[0];
	arctemp = "-$$$";
	arctemp[0] = tmpchr[0];
#endif
	arctemp2 = NULL;

#if	!UNIX
	/* avoid any case problems with arguments */

	for (n = 1; n < num; n++)	/* for each argument */
		upper(arg[n]);	/* convert it to uppercase */
#else
	/* avoid case problems with command options */
	upper(arg[1]);		/* convert to uppercase */
#endif

	for (n = 1; n < num; n++)
	  if (strlen(arg[n]) > (STRLEN - 16))
	  {
	    fprintf(stderr, "Too long argument: %s\n", arg[n]);
	    exit(235);
	  }

	/* create archive names, supplying defaults */
#if	UNIX
	if (!stat(arg[2],&sbuf)) {
		if ((sbuf.st_mode & S_IFMT) == S_IFDIR)
			makefnam(arg[2],".arc",arcname);
		else
			strcpy(arcname,arg[2]);
	} else
		makefnam(arg[2],".arc",arcname);
#else
	makefnam(arg[2], ".ARC", arcname);
#endif
	/* makefnam(".$$$",arcname,newname); */
	sprintf(newname, "%s.arc", arctemp);
	makefnam(".BAK", arcname, bakname);

	/* now scan the command and see what we are to do */

	for (a = arg[1]; *a; a++) {	/* scan the option flags */
#if	!_MTS
		if (index("AMUFDXEPLVTCR", *a)) {	/* if a known command */
#else
		if (index("AMUFDXEPLVTC", *a)) {
#endif
			if (opt)/* do we have one yet? */
				arcdie("Cannot mix %c and %c", opt, *a);
			opt = *a;	/* else remember it */
		} else if (*a == 'B')	/* retain backup copy */
			keepbak = 1;

		else if (*a == 'W')	/* suppress warnings */
			warn = 0;
#if	!DOS
		else if (*a == 'I')	/* image mode, no ASCII/EBCDIC x-late */
			image = !image;
#endif
#if	GEMDOS
		else if (*a == 'H')	/* pause before exit */
			hold = 1;
#endif

		else if (*a == 'N')	/* suppress notes and comments */
			note = 0;

		else if (*a == 'O')	/* overwrite file on extract */
			overlay = 1;

		else if (*a == 'G') {	/* garble */
			password = a + 1;
			while (*a)
				a++;
			a--;
#if	_MTS
			etoa(password, strlen(password));
#endif
		} else if (*a == 'S')	/* storage kludge */
			nocomp = 1;

		else if (*a == 'K')	/* special kludge */
			kludge = 1;

		else if (*a == 'Q')	/* use squashing */
			dosquash = 1;

		else if (*a == '-' || *a == '/')	/* UNIX and PC-DOS
							 * option markers */
			;

		else
			arcdie("%c is an unknown command", *a);
	}

	if (!opt)
		arcdie("I have nothing to do!");

	/* get the files list set up */

	lnum = num - 3;		/* initial length of list */
	lst = (char **) calloc((lnum==0) ? 1:lnum,
				 sizeof(char *));	/* initial list */
	for (n = 3; n < num; n++)
		lst[n - 3] = arg[n];

	for (n = 0; n < lnum;) {/* expand indirect references */
		if (*lst[n] == '@')
			expandlst(n);
#if	GEMDOS		/* redirect stdout from the desktop...*/
		else if (*lst[n] == '>') {
			arctemp2 = (++lst[n]);
			lst[n] = lst[--lnum];	/* delete this entry */
			if (arctemp2[0] == '>') {
				append = 1;
				arctemp2++;
			}
			else	append = 0;
		}
#endif
		else
			n++;
	}
#if	GEMDOS
	if (arctemp2)
		freopen(arctemp2,append ? "a" : "w",stdout);
#endif

	/* Allocate space to I/O buffers */

	if (!(pinbuf = malloc(MYBUF)))
		arcdie("Not enough memory for input buffer.");
	if (!(outbuf = malloc(MYBUF)))
		arcdie("Not enough memory for output buffer.");
	if (!(pakbuf = malloc(2L*MYBUF)))
		arcdie("Not enough memory for packing buffer.");
	outend = outbuf + MYBUF - 1;
	
	
	/* act on whatever action command was given */

	switch (opt) {		/* action depends on command */
	case 'A':		/* Add */
	case 'M':		/* Move */
	case 'U':		/* Update */
	case 'F':		/* Freshen */
		addarc(lnum, lst, (opt == 'M'), (opt == 'U'), (opt == 'F'));
		break;

	case 'D':		/* Delete */
		delarc(lnum, lst);
		break;

	case 'E':		/* Extract */
	case 'X':		/* eXtract */
	case 'P':		/* Print */
		extarc(lnum, lst, (opt == 'P'));
		break;

	case 'V':		/* Verbose list */
		bose = 1;
	case 'L':		/* List */
		lstarc(lnum, lst);
		break;

	case 'T':		/* Test */
		tstarc();
		break;

	case 'C':		/* Convert */
		cvtarc(lnum, lst);
		break;
#if	!_MTS
	case 'R':		/* Run */
		runarc(lnum, lst);
		break;
#endif
	default:
		arcdie("I don't know how to do %c yet!", opt);
	}
#if	GEMDOS
	if (hold)
		exitpause();
#endif
	return nerrs;
}
static	VOID
expandlst(n)			/* expand an indirect reference */
	int		n;	/* number of entry to expand */
{
	FILE	       *lf, *fopen();	/* list file, opener */
	char		buf[100];	/* input buffer */
	int		x;	/* index */
	char	       *p = lst[n] + 1; /* filename pointer */

	if (*p) {		/* use name if one was given */
		makefnam(p, ".CMD", buf);
		if (!(lf = fopen(buf, "r")))
			arcdie("Cannot read list of files in %s", buf);
	} else
		lf = stdin;	/* else use standard input */

	for (x = n + 1; x < lnum; x++)	/* drop reference from the list */
		lst[x - 1] = lst[x];
	lnum--;

	while (fscanf(lf, "%99s", buf) > 0) {	/* read in the list */
		if (!(lst =(char **)realloc(lst, (lnum + 1) * sizeof(char *))))
			arcdie("too many file references");

		lst[lnum] = malloc(strlen(buf) + 1);
		strcpy(lst[lnum], buf); /* save the name */
		lnum++;
	}

	if (lf != stdin)	/* avoid closing standard input */
		fclose(lf);
}
